File: backend/src/main/java/com/lsoft/aguapotable/controller/TarifaController.java
package com.lsoft.aguapotable.controller;

import com.lsoft.aguapotable.domain.entity.Tarifa;
import com.lsoft.aguapotable.dto.request.TarifaRequest;
import com.lsoft.aguapotable.service.TarifaService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/tarifas")
@RequiredArgsConstructor
@Tag(name = "Tarifas", description = "Gestión de tarifas escalonadas por tipo de suscriptor")
@SecurityRequirement(name = "bearerAuth")
public class TarifaController {

    private final TarifaService tarifaService;

    @GetMapping
    @Operation(summary = "Listar tarifas activas de la comunidad")
    public ResponseEntity<List<Tarifa>> listar(HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.ok(tarifaService.listarActivas(comunidadId));
    }

    @GetMapping("/{id}")
    @Operation(summary = "Obtener tarifa por ID")
    public ResponseEntity<Tarifa> getById(@PathVariable Long id, HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.ok(tarifaService.obtenerPorId(id, comunidadId));
    }

    @PostMapping
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN')")
    @Operation(summary = "Crear nueva tarifa con rangos escalonados")
    public ResponseEntity<Tarifa> crear(@Valid @RequestBody TarifaRequest dto,
                                         HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.status(201).body(tarifaService.crear(dto, comunidadId));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN')")
    @Operation(summary = "Actualizar tarifa y sus rangos")
    public ResponseEntity<Tarifa> actualizar(@PathVariable Long id,
                                              @Valid @RequestBody TarifaRequest dto,
                                              HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.ok(tarifaService.actualizar(id, dto, comunidadId));
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN')")
    @Operation(summary = "Desactivar tarifa")
    public ResponseEntity<Void> desactivar(@PathVariable Long id, HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        tarifaService.desactivar(id, comunidadId);
        return ResponseEntity.noContent().build();
    }
}

File: backend/src/tarifa/tarifa.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  ParseIntPipe,
  HttpCode,
  UseGuards,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { TarifaService } from './tarifa.service';
import { TarifaRequest } from './dto/tarifa.dto';
import { ComunidadId } from '../common/decorators';
import { RolesGuard, Roles } from '../auth/guards/roles.guard';

@ApiTags('Tarifas')
@ApiBearerAuth()
@Controller('api/v1/tarifas')
@UseGuards(AuthGuard('jwt'), RolesGuard)
export class TarifaController {
  constructor(private readonly tarifaService: TarifaService) {}

  @Get()
  @ApiOperation({ summary: 'Listar tarifas activas de la comunidad' })
  async listar(@ComunidadId() comunidadId: number) {
    return this.tarifaService.listarActivas(comunidadId);
  }

  @Get(':id')
  @ApiOperation({ summary: 'Obtener tarifa por ID' })
  async getById(
    @Param('id', ParseIntPipe) id: number,
    @ComunidadId() comunidadId: number,
  ) {
    return this.tarifaService.obtenerPorId(id, comunidadId);
  }

  @Post()
  @Roles('SUPER_ADMIN', 'ADMIN')
  @HttpCode(201)
  @ApiOperation({ summary: 'Crear nueva tarifa con rangos escalonados' })
  async crear(
    @Body() dto: TarifaRequest,
    @ComunidadId() comunidadId: number,
  ) {
    return this.tarifaService.crear(dto, comunidadId);
  }

  @Put(':id')
  @Roles('SUPER_ADMIN', 'ADMIN')
  @ApiOperation({ summary: 'Actualizar tarifa y sus rangos' })
  async actualizar(
    @Param('id', ParseIntPipe) id: number,
    @Body() dto: TarifaRequest,
    @ComunidadId() comunidadId: number,
  ) {
    return this.tarifaService.actualizar(id, dto, comunidadId);
  }

  @Delete(':id')
  @Roles('SUPER_ADMIN', 'ADMIN')
  @HttpCode(204)
  @ApiOperation({ summary: 'Desactivar tarifa' })
  async desactivar(
    @Param('id', ParseIntPipe) id: number,
    @ComunidadId() comunidadId: number,
  ) {
    return this.tarifaService.desactivar(id, comunidadId);
  }
}
